data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamScheduler.java

```diff
@@ -0,0 +1,85 @@
+package org.opensearch.dataprepper.plugins.mongo.stream;
+
+import org.opensearch.dataprepper.buffer.common.BufferAccumulator;
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSetManager;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.model.source.coordinator.enhanced.EnhancedSourceCoordinator;
+import org.opensearch.dataprepper.model.source.coordinator.enhanced.EnhancedSourcePartition;
+import org.opensearch.dataprepper.plugins.mongo.buffer.ExportRecordBufferWriter;
+import org.opensearch.dataprepper.plugins.mongo.buffer.RecordBufferWriter;
+import org.opensearch.dataprepper.plugins.mongo.configuration.MongoDBSourceConfig;
+import org.opensearch.dataprepper.plugins.mongo.converter.RecordConverter;
+import org.opensearch.dataprepper.plugins.mongo.coordination.partition.StreamPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+
+public class StreamScheduler implements Runnable {
+    private static final Logger LOG = LoggerFactory.getLogger(StreamScheduler.class);
+    private static final int DEFAULT_TAKE_LEASE_INTERVAL_MILLIS = 60_000;
```

Comment:
I would make this backoff smarter. 60 seconds is a long time to wait to grab another item to process. For shards in dynamodb, we do some custom backoff based on the number of shards this node has acquired, and if it acquired a shard previously (https://github.com/opensearch-project/data-prepper/blob/8760945c8091d1dd4e2bb8c66c5bb9f33f0aaa43/data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/StreamScheduler.java#L145). Something like this should work, at least to make it so 60 second wait isn't required when no partitions are acquired.

Reply:
For MongoDB/DocuentDB, there is only one partition to work which is done by the leader. If the leader crashes for some reason then this lease interval will kick in. I can make it better in follow up PR.